### PR TITLE
Enable more models to  inference based on LoRA

### PR DESCRIPTION
**Pull request by jeejeelee**
## MOTIVATION

For certain models such as ChatGLM3 and BaiChuan-7B in their `transformers` versions, the `QKVParallelLinear` layers are already combined. Consequently, when integrating LoRA support for them, utilizing the existing `QKVParallelLinearWithLora` is not feasible. The `dense_h_to_4h` layer of ChatGLM-3 operates in a similar fashion. New `linear` implementations are required to fulfill the lora integration needs for these models. Therefore, I have completed the following implementation:
- added a `RefQKVParallelLinear` identical to `QKVParallelLinear` to signify that QKV is already merged in their `transformer `versions. And based on `RefQKVParallelLinear`,implemented the corresponding LoRA layer ,named `QKVParallelLinearWithLora` .
- added a `RefMergedColumnParallelLinear` identical to `RefMergedColumnParallelLinear` to signify that `gated_up_proj` is already merged in their transformer versions.  And based on `RefMergedColumnParallelLinear`,implemented the corresponding LoRA layer ,named `RefMergedColumnParallelLinearWithLora` .
- completed the integration of LoRA support for ChatGLM3

### CODE REVIEW

File: csrc/punica/bgmv/bgmv_config.h [hunk 1]
```diff
@@ -16,34 +16,40 @@ void bgmv_kernel(out_T *__restrict__ Y, const in_T *__restrict__ X,
...
     f(in_T, out_T, W_T, narrow, 11008) \
     f(in_T, out_T, W_T, narrow, 12288) \
-    f(in_T, out_T, W_T, narrow, 13696) \
     f(in_T, out_T, W_T, narrow, 13824) \
+    f(in_T, out_T, W_T, narrow, 13696) \
     f(in_T, out_T, W_T, narrow, 14336) \
     f(in_T, out_T, W_T, narrow, 16384) \
     f(in_T, out_T, W_T, narrow, 20480) \
     f(in_T, out_T, W_T, narrow, 24576) \
+    f(in_T, out_T, W_T, narrow, 27392) \
```
[the remark this answers is not in the document]
Comment (jeejeelee) [reply csrc/punica/bgmv/bgmv_config.h:52]: ok , I will add these dimensions as soon as possibble

**Comment by jeejeelee** [comment]
@Yard1  Thanks for your review, I had added these new dimensions in `test_punica.py`, and the test result as follow:
```
3312 passed, 1872 skipped, 10 warnings
```

**Comment by Yard1** [comment]
Thanks - I am wondering if we can avoid adding the `Ref` layers. It should be possible to use the existing LoRA layers to represent already merged modules - one possibility is to change the logic based on the length of the value list in `packed_modules_mapping`. Ideally we'd not add new layers and instead modify existing ones.

**Comment by jeejeelee** [comment]
@Yard1 Thanks for your review and invaluable advise.
 I haved refactored code based on your suggestion.  I'm currently perplexed about how to properly name the new LoRA layers, so I still named them as "RefxxxWithLora".
As for the test, I'd like to confirm with you the above code refactoring before proceeding further.

File: vllm/lora/layers.py [hunk 2]
```diff
@@ -952,17 +1027,29 @@ def from_layer(
...
     for src_layer_type, lora_layer_type in supported_layer_types.items():
         if type(layer) is src_layer_type:  # pylint: disable=unidiomatic-typecheck
-            ret = lora_layer_type(layer)
+            if isinstance(lora_layer_type, tuple):
+                if len(packed_modules_list) > 1:
+                    ret = lora_layer_type[0](layer)
+                else:
+                    ret = lora_layer_type[1](layer)
+            else:
+                ret = lora_layer_type(layer)
```
Comment (Yard1) [inline vllm/lora/layers.py:1052]: this is too brittle. how about we add a class method to the lora layers that would act as a condition to be called here? eg. `can_replace_src_layer`

File: vllm/lora/layers.py [hunk 3]
```diff
@@ -657,6 +685,53 @@ def apply_weights(self, x: torch.Tensor,
         return output
 
 
+class RefQKVParallelLinearWithLora(ColumnParallelLinearWithLoRA):
```
Comment (Yard1) [inline vllm/lora/layers.py:688]: Let's name this `QKVParallelLinearWithLora` and rename current `QKVParallelLinearWithLora` to `MergedQKVParallelLinearWithLora` (to be consistent with `ColumnParallelLinear`)

let's also add docstrings to both new layers

File: vllm/lora/layers.py [hunk 4]
```diff
@@ -488,6 +488,34 @@ def apply_weights(self, x: torch.Tensor,
         return output
 
 
+class RefMergedColumnParallelLinearWithLoRA(ColumnParallelLinearWithLoRA):
```
Comment (Yard1) [inline vllm/lora/layers.py:491]: we already have `ColumnParallelLinearWithLoRA`, how is this different?

File: vllm/lora/models.py [hunk 5]
```diff
@@ -508,8 +509,8 @@ def _match_target_modules(self, module_name: str):
     def _register_packed_modules(self, module_full_name: str) -> None:
         parts = module_full_name.split(".")
         module_name = parts[-1]
-        replacements = self.packed_modules_mapping.get(module_name)
-        if not replacements:
+        replacements = self.packed_modules_mapping.get(module_name, [])
+        if len(replacements) <= 1:
```
Comment (Yard1) [inline vllm/lora/models.py:513]: this change is not obvious. can you add a comment explaining what happens here?

**Review by Yard1 — CHANGES_REQUESTED** [review]

**Comment by yuanmeng1120** [comment]
您好，这个pr 改后，baichuan2 _13B _chat加lora 和不加的结果是一样的，都是回复我是百川大模型，但是lora 的模型是做了身份矫正的，回复的答案不对，感觉lora 没起作用。

[
WechatIMG1180
](url)

**Comment by jeejeelee** [comment]
> 您好，这个pr 改后，baichuan2 _13B _chat加lora 和不加的结果是一样的，都是回复我是百川大模型，但是lora 的模型是做了身份矫正的，回复的答案不对，感觉lora 没起作用。
> 
> [ WechatIMG1180 ](url)
感谢关注，
1. 请问你修改了vllm的baichuan的网络结构了吗
2. 有基于tansformers验证过lora吗

**Comment by jeejeelee** [comment]
@yuanmeng1120  我已完成百川相关模型支持lora的相关代码书写并使用baichuan-7B测试通过，您可以再次尝试下

**Comment by jeejeelee** [comment]
@Yard1  Can you take a little time to review this PR again, ths~

**Comment by yuanmeng1120** [comment]
@jeejeelee 您好，我刚开始是在BaichuanForCausalLM 和BaiChuanForCausalLM 的两个类中改的，没有在base里改动，只有  "gata_down_proj": [
    #         "gate_proj",
    #         "down_proj",
    #     ], 这里不一样， 因为我看baichuan结构是先down, 后up , 其他的一样，是下面问题：
    lora/layers.py", line 595, in set_lora
    lora_b_q = lora_b[0][:, self.q_proj_shard_size *
TypeError: list indices must be integers or slices, not tuple
然后又按照您提的pr 改了后还是下面问题：
lora/layers.py", line 595, in set_lora
    lora_b_q = lora_b[0][:, self.q_proj_shard_size *
TypeError: list indices must be integers or slices, not tuple
我再调试下修改下结构试试哈，非常感谢


> @yuanmeng1120 我已完成百川相关模型支持lora的相关代码书写并使用baichuan-7B测试通过，您可以再次尝试下



> @yuanmeng1120 我已完成百川相关模型支持lora的相关代码书写并使用baichuan-7B测试通过，您可以再次尝试下

**Comment by jeejeelee** [comment]
@yuanmeng1120  Hi,如果您是拉了我的[分支](https://github.com/jeejeelee/vllm/tree/modify-kqv-merged-linear),再重新编译即可，应该不需要再修改百川模型的代码了。如果只是修改vllm主分支的代码的话，可以参考本次PR的[修改内容](https://github.com/vllm-project/vllm/pull/3382/files)来进行相关的修改

**Comment by yuanmeng1120** [comment]
> @yuanmeng1120 你好，如果你拉了我的[分支](https://github.com/jeejeelee/vllm/tree/modify-kqv-merged-linear)，再重新编译即可，不需要再修改百川模型的代码了。如果只是修改vllm主分支的代码的话，可以参考本次PR的[内容](https://github.com/vllm-project/vllm/pull/3382/files)来进行相关的修改

好的，感谢，我之前没拉pr， 改的源码，我拉下pr 试下

**Comment by yuanmeng1120** [comment]
> @yuanmeng1120 你好，如果你拉了我的[分支](https://github.com/jeejeelee/vllm/tree/modify-kqv-merged-linear)，再重新编译即可，不需要再修改百川模型的代码了。如果只是修改vllm主分支的代码的话，可以参考本次PR的[内容](https://github.com/vllm-project/vllm/pull/3382/files)来进行相关的修改

@jeejeelee 您好，拉了您的pr代码，然后报了这个错误：

**Comment by jeejeelee** [comment]
@yuanmeng1120  Hi,其实这个问题我在跑chatglm3时，按照TP=4的时候也会遇到，在punica的config.h添加的话，的确是因为不满足被64整除的问题导致编译期断言失败。若是你也跑的chatglm3的话，建议TP=1或者2再尝试下

**Comment by yuanmeng1120** [comment]
> @yuanmeng1120 Hi,其实这个问题我在跑chatglm3时，按照TP=4的时候遇到，在punica的config.h添加的话，确实是因为不满足被64整除的问题导致编译期断言失败。如果你也跑的chatglm3的话，TP=1或者2再尝试下



> @yuanmeng1120 你好，其实这个问题我在跑chatglm3时，按照TP=4的时候遇到，在punica的config.h添加的话，确实是因为不满足被64整除的问题导致编译期断言失败。如果你也跑的chatglm3的话，TP=1或者2再尝试下
@jeejeelee 可以了，确实是这个问题，非常感谢 ， 我试了2 不行， 1可以的

**Comment by Yard1** [comment]
@jeejeelee can you address the comments I left before? Thanks!

**Comment by nlp-learner** [comment]
support chatglm2?

**Comment by jeejeelee** [comment]
> 是否需要更改以下代码？ if 32000 < self.base_layer.vocab_size > 33024: raise ValueError("When using LoRA, vocab size must be " "32000 >= vocab_size <= 33024")

就个人的使用经验，不对`llm_head`添加lora，所以也不需要修改

**Comment by jeejeelee** [comment]
@Yard1 I apologize for bothering you. My goal is to merge this PR before the v0.34 release, if feasible. Your feedback on the above comment would be greatly appreciated.

**Review by Yard1 — APPROVED** [review]
Thanks, looks good. Let's merge once CI passses.

**Comment by jeejeelee** [comment]
@Yard1  Thank you for your great work.

**Pull request merged**